internal/xds/clients/xdsclient/ads_stream.go

```diff
@@ -280,53 +288,44 @@ func (s *adsStreamImpl) send(ctx context.Context) {
 				stream = nil
 				continue
 			}
-		case req, ok := <-s.requestCh.Get():
-			if !ok {
-				return
+		case <-s.notifySender:
+			// If there's no stream yet, skip the request. This request will be resent
+			// when a new stream is created. If no stream is created, the watcher will
+			// timeout (same as server not sending response back).
+			if stream == nil {
+				continue
 			}
-			s.requestCh.Load()
 
-			typ := req.(ResourceType)
-			if err := s.sendNew(stream, typ); err != nil {
+			// Resetting the pendingRequests slice to nil works for both cases:
+			// - When we successfully sends the requests out on the wire.
+			// - When sending fails. This can happen only when the stream fails,
+			//   and in this case, we rely on the `sendExisting` to send out
+			//   requests for all subscriptions when the stream is recreated.
+			s.mu.Lock()
+			if err := s.sendNewLocked(stream, s.pendingRequests); err != nil {
 				stream = nil
-				continue
 			}
+			s.pendingRequests = nil
+			s.mu.Unlock()
 		}
```

Comment:
for my education, there is no need for a `default` condition here?

Reply:
The `default` condition works very differently in a `select` statement compared to how it works in a `switch` statement. If we had a `default` case here, we would end up busy looping in the `for` loop because for most of the time the other reads from the three channels are going to be blocked.